File: ai-services/me.ai.service.js
/**
 * ai-services/me.ai.service.js
 *
 * Pure-JS port of me_ai_service.py.
 *
 * Replaces the previous spawn("python", ...) wrapper. Same exported
 * names + return shapes (`rerankForYouWithPythonAI`, `annotateWorkedOnWithAI`)
 * so modules/me consumes it unchanged.
 *
 * Model identifier bumped to `js-heuristic-v1` so analytics dashboards
 * can see the cutover. All scoring weights kept identical to the
 * Python version so ranked output is bit-for-bit equivalent.
 */

// ── Helpers ────────────────────────────────────────────────────

function parseDate(value) {
  if (value == null || value === "") return null;
  if (typeof value === "object" && value.$date) value = value.$date;
  if (value instanceof Date) return value;
  const t = String(value).trim();
  if (!t) return null;
  const d = new Date(t.endsWith("Z") ? t : t);
  return Number.isNaN(d.getTime()) ? null : d;
}

function parseId(value) {
  if (value == null) return null;
  if (typeof value === "object" && value.$oid) return String(value.$oid);
  return String(value);
}

const clamp = (x, lo = 0, hi = 100) => Math.max(lo, Math.min(hi, x));
const round2 = (n) => Math.round(n * 100) / 100;

function bump(map, key) {
  map.set(key, (map.get(key) || 0) + 1);
}

function recencyBonus(updatedAt, now) {
  const dt = parseDate(updatedAt);
  if (!dt) return 0;
  const hours = Math.max((now - dt) / 3_600_000, 0);
  if (hours <= 24) return 12;
  if (hours <= 72) return 7;
  if (hours <= 168) return 3;
  return 0;
}

// ── AI-6.1 basic relevance ─────────────────────────────────────

function basicRelevanceScore(task, now) {
  const priority = String(task.priority || "Low");
  const status = String(task.status || "Todo");
  const priorityMap = { Urgent: 40, High: 30, Medium: 18, Low: 8 };
  const statusMap = { InProgress: 24, Todo: 14, Done: -35 };
  const score =
    (priorityMap[priority] ?? 8) +
    (statusMap[status] ?? 8) +
    recencyBonus(task.updatedAt, now);
  return clamp(score + 20);
}

// ── AI-6.4 time urgency ────────────────────────────────────────

function timeUrgencyScore(task, now) {
  const due = parseDate(task.dueDate);
  if (!due) return 5;
  const diffDays = (due - now) / 86_400_000;
  if (diffDays < 0) return 100;
  if (diffDays <= 1) return 90;
  if (diffDays <= 3) return 75;
  if (diffDays <= 7) return 55;
  if (diffDays <= 14) return 35;
  return 15;
}

// ── User profile (used by AI-6.2 / 6.3 / 6.5) ──────────────────

function buildUserProfile(tasks) {
  const byType = new Map();
  const byPriority = new Map();
  const byLabel = new Map();
  const bySpace = new Map();
  const completionSamples = [];

  for (const t of tasks) {
    const ttype = String(t.type || "Task");
    const pr = String(t.priority || "Low");
    const status = String(t.status || "Todo");
    const spaceId = parseId(t.spaceId);

    bump(byType, ttype);
    bump(byPriority, pr);
    if (spaceId) bump(bySpace, spaceId);

    for (const lb of t.labels || []) bump(byLabel, String(lb).toLowerCase());

    if (status === "Done") {
      const created = parseDate(t.createdAt);
      const updated = parseDate(t.updatedAt);
      if (created && updated && updated >= created) {
        completionSamples.push((updated - created) / 3_600_000);
      }
    }
  }
  return { byType, byPriority, byLabel, bySpace, completionSamples };
}

// ── AI-6.2 behavior ────────────────────────────────────────────

function userBehaviorScore(task, profile) {
  const { byType, byPriority, byLabel } = profile;
  const totalType = sumMap(byType) || 1;
  const totalPriority = sumMap(byPriority) || 1;
  const totalLabel = sumMap(byLabel) || 1;

  const ttype = String(task.type || "Task");
  const pr = String(task.priority || "Low");

  let score = 35 * ((byType.get(ttype) || 0) / totalType);
  score += 25 * ((byPriority.get(pr) || 0) / totalPriority);

  const labels = (task.labels || []).map((x) => String(x).toLowerCase());
  if (labels.length) {
    const labelFit = Math.max(
      ...labels.map((lb) => (byLabel.get(lb) || 0) / totalLabel),
    );
    score += 40 * labelFit;
  } else {
    score += 5;
  }
  return clamp(score);
}

// ── AI-6.3 collaborative ───────────────────────────────────────

function collaborativeFilterScore(task, teamTasks, userProfile) {
  if (!teamTasks?.length) return 10;

  const teamType = new Map();
  const teamLabel = new Map();
  const teamSpace = new Map();
  for (const t of teamTasks) {
    bump(teamType, String(t.type || "Task"));
    const sid = parseId(t.spaceId);
    if (sid) bump(teamSpace, sid);
    for (const lb of t.labels || []) bump(teamLabel, String(lb).toLowerCase());
  }
  const totalTeamType = sumMap(teamType) || 1;
  const totalTeamLabel = sumMap(teamLabel) || 1;
  const totalTeamSpace = sumMap(teamSpace) || 1;

  const ttype = String(task.type || "Task");
  const spaceId = parseId(task.spaceId);
  const labels = (task.labels || []).map((x) => String(x).toLowerCase());

  const typeFit = (teamType.get(ttype) || 0) / totalTeamType;
  const spaceFit = spaceId
    ? (teamSpace.get(spaceId) || 0) / totalTeamSpace
    : 0;
  const labelFit = labels.length
    ? Math.max(
        ...labels.map((lb) => (teamLabel.get(lb) || 0) / totalTeamLabel),
      )
    : 0;

  // Novelty: prefer team signal that user does less often
  const userTypeCount = userProfile.byType.get(ttype) || 0;
  const novelty = 1 / (1 + userTypeCount);

  return clamp(45 * typeFit + 30 * labelFit + 20 * spaceFit + 20 * novelty);
}

// ── AI-6.5 project/space affinity ──────────────────────────────

function projectAffinityScore(task, profile) {
  const totalSpace = sumMap(profile.bySpace) || 1;
  const sid = parseId(task.spaceId);
  const spaceFit = sid ? (profile.bySpace.get(sid) || 0) / totalSpace : 0;
  const parentBonus = task.parentTaskId ? 12 : 0;
  return clamp(20 + 70 * spaceFit + parentBonus);
}

function explainComponents(components) {
  const labels = {
    relevance: "AI-6.1 relevance",
    behavior: "AI-6.2 behavior",
    collab: "AI-6.3 collaborative",
    urgency: "AI-6.4 urgency",
    affinity: "AI-6.5 affinity",
  };
  const top = Object.entries(components)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 2);
  return top.map(([k]) => labels[k]).join(", ");
}

// ── Public API (preserves the previous wrapper's exports) ──────

export async function rerankForYouWithPythonAI({
  scored = [],
  limit = 15,
  userHistory = [],
  teamHistory = [],
} = {}) {
  const now = new Date();
  const userProfile = buildUserProfile(userHistory);
  const cap = Number(limit) || 15;

  const ranked = scored.map((row) => {
    const task = row.task || {};
    const baseRule = Number(row.score) || 0;

    const cRelevance = basicRelevanceScore(task, now);
    const cBehavior = userBehaviorScore(task, userProfile);
    const cCollab = collaborativeFilterScore(task, teamHistory, userProfile);
    const cUrgency = timeUrgencyScore(task, now);
    const cAffinity = projectAffinityScore(task, userProfile);

    const aiScore =
      0.22 * cRelevance +
      0.20 * cBehavior +
      0.18 * cCollab +
      0.25 * cUrgency +
      0.15 * cAffinity;

    const finalScore = round2(baseRule * 0.35 + aiScore * 0.65);
    const components = {
      relevance: round2(cRelevance),
      behavior: round2(cBehavior),
      collab: round2(cCollab),
      urgency: round2(cUrgency),
      affinity: round2(cAffinity),
    };

    return {
      ...row,
      score: finalScore,
      aiScore: round2(aiScore),
      aiReason: explainComponents(components),
      aiBreakdown: components,
    };
  });

  ranked.sort((a, b) => (Number(b.score) || 0) - (Number(a.score) || 0));

  return {
    aiUsed: true,
    aiModel: "js-heuristic-v1",
    aiFallbackReason: null,
    items: ranked.slice(0, cap),
  };
}

// Legacy alias preserved so existing callers don't break.
export async function rerankForYouWithAI(args) {
  return rerankForYouWithPythonAI(args);
}

export async function annotateWorkedOnWithAI({ items = [] } = {}) {
  const now = new Date();
  const notes = [];
  // Match Python behavior: cap at first 60 items.
  for (const t of items.slice(0, 60)) {
    const taskId = parseId(t._id);
    const status = String(t.status || "Todo");
    const urgency = timeUrgencyScore(t, now);
    const relevance = basicRelevanceScore(t, now);
    const score = round2(0.55 * urgency + 0.45 * relevance);

    let note;
    if (status === "Done") note = "Completed recently; keep for context.";
    else if (urgency >= 75) note = "High urgency due date; prioritize next.";
    else if (relevance >= 65)
      note = "Strong personal relevance based on activity.";
    else note = "Moderate relevance; schedule when free.";

    notes.push({ id: taskId, score, note });
  }

  // Existing consumer expects `notesById` as a Map.
  const notesById = new Map();
  for (const n of notes) {
    if (!n?.id) continue;
    notesById.set(String(n.id), { aiScore: n.score, aiNote: n.note });
  }

  return {
    aiUsed: true,
    aiModel: "js-heuristic-v1",
    aiFallbackReason: null,
    notesById,
  };
}

// ── helpers ────────────────────────────────────────────────────
function sumMap(m) {
  let s = 0;
  for (const v of m.values()) s += v;
  return s;
}

File: ai-services/me_ai_service.py
import json
import math
import sys
from collections import Counter, defaultdict
from datetime import datetime, timezone


def parse_date(value):
    if not value:
        return None
    if isinstance(value, dict) and "$date" in value:
        value = value["$date"]
    if isinstance(value, datetime):
        return value
    text = str(value).strip()
    if not text:
        return None
    if text.endswith("Z"):
        text = text[:-1] + "+00:00"
    try:
        dt = datetime.fromisoformat(text)
    except Exception:
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt


def parse_id(value):
    if isinstance(value, dict) and "$oid" in value:
        return str(value["$oid"])
    if value is None:
        return None
    return str(value)


def clamp(x, lo=0.0, hi=100.0):
    return max(lo, min(hi, x))


def recency_bonus(updated_at, now):
    dt = parse_date(updated_at)
    if not dt:
        return 0.0
    hours = max((now - dt).total_seconds() / 3600.0, 0.0)
    if hours <= 24:
        return 12.0
    if hours <= 72:
        return 7.0
    if hours <= 168:
        return 3.0
    return 0.0


# AI-6.1
def basic_relevance_score(task, now):
    priority = str(task.get("priority", "Low"))
    status = str(task.get("status", "Todo"))

    priority_map = {"Urgent": 40, "High": 30, "Medium": 18, "Low": 8}
    status_map = {"InProgress": 24, "Todo": 14, "Done": -35}

    score = priority_map.get(priority, 8) + status_map.get(status, 8)
    score += recency_bonus(task.get("updatedAt"), now)

    return clamp(score + 20)


# AI-6.4
def time_urgency_score(task, now):
    due = parse_date(task.get("dueDate"))
    if not due:
        return 5.0

    diff_days = (due - now).total_seconds() / 86400.0
    if diff_days < 0:
        return 100.0
    if diff_days <= 1:
        return 90.0
    if diff_days <= 3:
        return 75.0
    if diff_days <= 7:
        return 55.0
    if diff_days <= 14:
        return 35.0
    return 15.0


def build_user_profile(tasks):
    by_type = Counter()
    by_priority = Counter()
    by_label = Counter()
    by_space = Counter()

    completion_samples = []

    for t in tasks:
        ttype = str(t.get("type", "Task"))
        pr = str(t.get("priority", "Low"))
        status = str(t.get("status", "Todo"))
        space_id = parse_id(t.get("spaceId"))

        by_type[ttype] += 1
        by_priority[pr] += 1
        if space_id:
            by_space[space_id] += 1

        for lb in t.get("labels") or []:
            by_label[str(lb).lower()] += 1

        if status == "Done":
            created = parse_date(t.get("createdAt"))
            updated = parse_date(t.get("updatedAt"))
            if created and updated and updated >= created:
                completion_samples.append((updated - created).total_seconds() / 3600.0)

    return {
        "by_type": by_type,
        "by_priority": by_priority,
        "by_label": by_label,
        "by_space": by_space,
        "completion_samples": completion_samples,
    }


# AI-6.2
def user_behavior_score(task, profile):
    score = 0.0
    by_type = profile["by_type"]
    by_priority = profile["by_priority"]
    by_label = profile["by_label"]

    total_type = sum(by_type.values()) or 1
    total_priority = sum(by_priority.values()) or 1
    total_label = sum(by_label.values()) or 1

    ttype = str(task.get("type", "Task"))
    pr = str(task.get("priority", "Low"))

    score += 35.0 * (by_type.get(ttype, 0) / total_type)
    score += 25.0 * (by_priority.get(pr, 0) / total_priority)

    labels = [str(x).lower() for x in (task.get("labels") or [])]
    if labels:
        label_fit = max((by_label.get(lb, 0) / total_label) for lb in labels)
        score += 40.0 * label_fit
    else:
        score += 5.0

    return clamp(score)


# AI-6.3
def collaborative_filter_score(task, team_tasks, user_profile):
    if not team_tasks:
        return 10.0

    team_type = Counter()
    team_label = Counter()
    team_space = Counter()
    for t in team_tasks:
        team_type[str(t.get("type", "Task"))] += 1
        sid = parse_id(t.get("spaceId"))
        if sid:
            team_space[sid] += 1
        for lb in t.get("labels") or []:
            team_label[str(lb).lower()] += 1

    total_team_type = sum(team_type.values()) or 1
    total_team_label = sum(team_label.values()) or 1
    total_team_space = sum(team_space.values()) or 1

    ttype = str(task.get("type", "Task"))
    space_id = parse_id(task.get("spaceId"))
    labels = [str(x).lower() for x in (task.get("labels") or [])]

    type_fit = team_type.get(ttype, 0) / total_team_type
    space_fit = (team_space.get(space_id, 0) / total_team_space) if space_id else 0.0
    label_fit = max((team_label.get(lb, 0) / total_team_label) for lb in labels) if labels else 0.0

    # Novelty: prefer team signal that user does less often.
    user_type_count = user_profile["by_type"].get(ttype, 0)
    novelty = 1.0 / (1.0 + user_type_count)

    score = 45.0 * type_fit + 30.0 * label_fit + 20.0 * space_fit + 20.0 * novelty
    return clamp(score)


# AI-6.5
def project_affinity_score(task, profile):
    by_space = profile["by_space"]
    total_space = sum(by_space.values()) or 1
    sid = parse_id(task.get("spaceId"))
    space_fit = (by_space.get(sid, 0) / total_space) if sid else 0.0

    # Mild bonus if task has parent epic/story (deeper project context)
    has_parent = task.get("parentTaskId") is not None
    parent_bonus = 12.0 if has_parent else 0.0

    return clamp(20.0 + 70.0 * space_fit + parent_bonus)


def explain_components(components):
    labels = {
        "relevance": "AI-6.1 relevance",
        "behavior": "AI-6.2 behavior",
        "collab": "AI-6.3 collaborative",
        "urgency": "AI-6.4 urgency",
        "affinity": "AI-6.5 affinity",
    }
    top = sorted(components.items(), key=lambda kv: kv[1], reverse=True)[:2]
    return ", ".join(labels[k] for k, _ in top)


def for_you_v2(payload):
    scored = payload.get("scored", []) or []
    limit = int(payload.get("limit", 15))
    user_history = payload.get("userHistory", []) or []
    team_history = payload.get("teamHistory", []) or []

    now = datetime.now(timezone.utc)
    user_profile = build_user_profile(user_history)

    ranked = []
    for row in scored:
        task = row.get("task", {}) or {}
        base_rule = float(row.get("score", 0) or 0)

        c_relevance = basic_relevance_score(task, now)
        c_behavior = user_behavior_score(task, user_profile)
        c_collab = collaborative_filter_score(task, team_history, user_profile)
        c_urgency = time_urgency_score(task, now)
        c_affinity = project_affinity_score(task, user_profile)

        # Weighted blend of AI-6.1..AI-6.5
        ai_score = (
            0.22 * c_relevance
            + 0.20 * c_behavior
            + 0.18 * c_collab
            + 0.25 * c_urgency
            + 0.15 * c_affinity
        )

        final_score = round(base_rule * 0.35 + ai_score * 0.65, 2)
        components = {
            "relevance": round(c_relevance, 2),
            "behavior": round(c_behavior, 2),
            "collab": round(c_collab, 2),
            "urgency": round(c_urgency, 2),
            "affinity": round(c_affinity, 2),
        }

        out = dict(row)
        out["score"] = final_score
        out["aiScore"] = round(ai_score, 2)
        out["aiReason"] = explain_components(components)
        out["aiBreakdown"] = components
        ranked.append(out)

    ranked.sort(key=lambda x: float(x.get("score", 0) or 0), reverse=True)

    return {
        "aiUsed": True,
        "aiModel": "python-heuristic-v1",
        "aiFallbackReason": None,
        "items": ranked[:limit],
    }


def worked_on(payload):
    items = payload.get("items", []) or []
    now = datetime.now(timezone.utc)
    notes = []
    for t in items[:60]:
        task_id = parse_id(t.get("_id"))
        status = str(t.get("status", "Todo"))
        urgency = time_urgency_score(t, now)
        relevance = basic_relevance_score(t, now)
        score = round(0.55 * urgency + 0.45 * relevance, 2)

        if status == "Done":
            note = "Completed recently; keep for context."
        elif urgency >= 75:
            note = "High urgency due date; prioritize next."
        elif relevance >= 65:
            note = "Strong personal relevance based on activity."
        else:
            note = "Moderate relevance; schedule when free."

        notes.append({"id": task_id, "score": score, "note": note})

    return {
        "aiUsed": True,
        "aiModel": "python-heuristic-v1",
        "aiFallbackReason": None,
        "notes": notes,
    }


def for_you_legacy(payload):
    # Backward command alias
    return for_you_v2(payload)


def main():
    command = sys.argv[1] if len(sys.argv) > 1 else ""
    raw = sys.stdin.read() or "{}"
    try:
        payload = json.loads(raw)
    except Exception:
        payload = {}

    try:
        if command == "for_you_v2":
            out = for_you_v2(payload)
        elif command == "for_you":
            out = for_you_legacy(payload)
        elif command == "worked_on":
            out = worked_on(payload)
        else:
            out = {
                "aiUsed": False,
                "aiModel": None,
                "aiFallbackReason": "Unknown command",
            }
    except Exception as exc:
        out = {
            "aiUsed": False,
            "aiModel": "python-heuristic-v1",
            "aiFallbackReason": str(exc),
        }

    sys.stdout.write(json.dumps(out))


if __name__ == "__main__":
    main()
